rp2040-hal/src/usb.rs

```diff
@@ -50,6 +50,85 @@ use usb_device::{
     Result as UsbResult, UsbDirection, UsbError,
 };
 
+unsafe fn errata5_workaround_finish((prev_pads, prev_io_ctrls, start_ts): (u32, u32, u32)) -> bool {
+    let pac = crate::pac::Peripherals::steal();
+    let now = pac.TIMER.timerawl.read().bits();
+    if now.wrapping_sub(start_ts) < 1_000 {
+        return false;
+    }
+    let pads = &pac.PADS_BANK0.gpio[15];
+    let io = &pac.IO_BANK0.gpio[15];
+
+    // Should think we are connected now, tight loop just in case
+    while pac
+        .USBCTRL_REGS
+        .sie_status
+        .read()
+        .connected()
+        .bit_is_clear()
+    {
+        cortex_m::asm::nop();
+    }
+
+    // Switch back to USB phy
+    pac.USBCTRL_REGS
+        .usb_muxing
+        .write(|w| w.to_phy().set_bit().softcon().set_bit());
+
+    // Get rid of DP pullup override
+    pac.USBCTRL_REGS
+        .usbphy_direct_override
+        .write(|w| w.dp_pullup_en_override_en().set_bit());
+
+    // Finally, restore the gpio ctrl value back to GPIO15
+    io.gpio_ctrl.write(|w| w.bits(prev_io_ctrls));
+    // Restore the pad ctrl value
+    pads.write(|w| w.bits(prev_pads));
+    true
+}
+unsafe fn errata5_workaround_start() -> (u32, u32, u32) {
+    let pac = crate::pac::Peripherals::steal();
+    let pads = &pac.PADS_BANK0.gpio[15];
+    let io = &pac.IO_BANK0.gpio[15];
+
+    // backup io ctrl & pad ctrl
+    let prev_pads = pads.read().bits();
+    let prev_io_ctrls = io.gpio_ctrl.read().bits();
+
+    // Enable bus keep and force pin to tristate, so USB DP muxing doesn't affect
+    // pin state
+    pads.modify(|_, w| w.pue().set_bit().pde().set_bit());
+    io.gpio_ctrl.modify(|_, w| w.oeover().disable());
+
+    // Select function 8 (USB debug muxing) without disturbing other controls
+    io.gpio_ctrl.modify(|_, w| w.funcsel().bits(8));
+
+    // J state is a differential 1 for a full speed device so
+    // DP = 1 and DM = 0. Don't actually need to set DM low as it
+    // is already gated assuming it isn't funcseld.
+    io.gpio_ctrl.modify(|_, w| w.inover().high());
+
+    // Force PHY pull up to stay before switching away from the phy
+    pac.USBCTRL_REGS
+        .usbphy_direct
+        .write(|w| w.dp_pullup_en().set_bit());
+    pac.USBCTRL_REGS
+        .usbphy_direct_override
+        .write(|w| w.dp_pullup_en_override_en().set_bit());
```

Comment:
The SDK code uses hw_set_alias here, so you should be preserving the other bits in the register by calling modify() instead of write()